[Docs] Update site2/docs/functions-overview.md (#4507)
apache/pulsar: site2/docs/functions-overview.md

* consume messages from one or more Pulsar topics,
* apply a user-supplied processing logic to each message,
* publish the results of the computation to another topic.

> The following is an example of Pulsar Functions for Java (using the [native interface](functions-api.md#java-native-functions)).

Reviewer: ```suggestion
The following is an example of a Pulsar Function written in Java (using the [native interface](functions-api.md#java-native-functions)).
```

}
```

> The following is an example of Pulsar Functions for Python (using the [native interface](functions-api.md#python-native-functions)).

Reviewer: ```suggestion
The following is an example of a Pulsar Function written in Python (using the [native interface](functions-api.md#python-native-functions)).
```

```python
def process(input):
    return "{0}!".format(input)
```

> The following is an example of Pulsar Functions for Go.

Reviewer: ```suggestion
The following is an example of a Pulsar Functions written in Go.
```

```
import (
	"fmt"
	"context"

	"github.com/apache/pulsar/pulsar-function-go/pf"
)

func HandleRequest(ctx context.Context, in []byte) error {
	fmt.Println(string(in) + "!")
	return nil
}

func main() {
	pf.Start(HandleRequest)
}
```

> Functions are executed each time a message is published to the input topic. For example, if a function is listening to the `tweet-stream` topic, the function runs each time when a message is published to that topic.

Reviewer: ```suggestion
A Pulsar Function is executed each time a message is published to it's input topic. For example, if a function has an input topic called `tweet-stream`, the function runs each time a message is published to `tweet-stream`.
```

## Goals

> With Pulsar Functions, you can create complex processing logic without deploying a separate neighboring system (such as [Apache Storm](http://storm.apache.org/), [Apache Heron](https://apache.github.io/incubator-heron), [Apache Flink](https://flink.apache.org/)). Pulsar Functions are computing infrastructure of Pulsar messaging system. The core goal is tied to a series of other goals:

Reviewer: Correct usage of Pulsar Functions here to reference the framework

## Fully Qualified Function Name (FQFN)

> Each Pulsar Functions has a **Fully Qualified Function Name** (FQFN) that consists of three elements: the function tenant, namespace, and function name. FQFN looks like this:

Reviewer: why do we need the change "Function" to be plural? That doesn't seem correct

```http
tenant/namespace/name
```

FQFNs enable you to create multiple functions with the same name provided that they are in different namespaces.

## Configuration

> You can configure Pulsar Functions in the following ways:

Reviewer: ```suggestion
You can configure a Pulsar Function in the following ways:
```

output: persistent://public/default/test_result
```

> You can specify some function attributes with CLI, and specify other attributes in YAML configuration.

Reviewer: ```suggestion
You can specify function attributes via CLI arguments or in a configuration file in YAML format.
```

## Supported languages

Currently, you can write Pulsar Functions in [Java](functions-api.md#functions-for-java), [Python](functions-api.md#functions-for-python), and [Go](functions-api.md#functions-for-go). Support for additional languages is coming soon.

## Pulsar Functions API

Pulsar Functions API enables you to create processing logic that is:

* Type safe. Pulsar Functions can process raw bytes or more complex, application-specific types.
* Based on SerDe (**Ser**ialization/**De**serialization). A variety of types are supported "out of the box" but you can also create your own custom SerDe logic.

### Function context

> Each Pulsar Functions created using [Pulsar Functions SDK](#the-pulsar-functions-sdk) has access to a context object that both provides:

Reviewer: ```suggestion
Each Pulsar Function created using [Pulsar Functions SDK](#the-pulsar-functions-sdk) has access to a context object that both provides:
```

### Language-native functions

> Both Java and Python support writing "native" functions, which mean Pulsar Functions with no dependencies.

Reviewer: ```suggestion
"Native"  functions are supported in Java and Python,  which means a Pulsar Function can have no dependencies.
```

The benefit of native functions is that they do not have any dependencies beyond what's already available in Java/Python "out of the box." The downside is that they do not provide access to the function [context](#function-context), which is necessary for a variety of functionalities, including [logging](#logging), [user configuration](#user-configuration), and more.

## Pulsar Functions SDK

> To enable Pulsar Functions to access to a [context object](#function-context), you can use **Pulsar Functions SDK**, available for [Java](functions-api.md#functions-for-java), [Python](functions-api.md#functions-for-python), and [Go](functions-api.md#functions-for-go).

Reviewer: `
```suggestion
To enable a Pulsar Function to access to a [context object](#function-context), you can use **Pulsar Functions SDK**, available for [Java](functions-api.md#functions-for-java), [Python](functions-api.md#functions-for-python), and [Go](functions-api.md#functions-for-go).
```

        log.info("Function tenant/namespace/name: {0}/{1}/{2}".format(function_tenant, function_namespace, function_name))
```

### Go

The following is a Go function example that uses information about its context.

```
import (
	"context"
	"fmt"

    "github.com/apache/pulsar/pulsar-function-go/log"
	"github.com/apache/pulsar/pulsar-function-go/pf"
)

func contextFunc(ctx context.Context) {
	if fc, ok := pf.FromContext(ctx); ok {
		tenant := fc.GetFuncTenant()
		namespace := fc.GetFuncNamespace()
		name := fc.GetFuncName()
		log.Info("Function tenant/namespace/name: %s/%s/%s\n", tenant, namespace, name)
	}
}

func main() {
	pf.Start(contextFunc)
}

```

## Deployment

> Pulsar Functions support a variety of deployment options. You can run Pulsar Functions in the following ways.

Reviewer: ```suggestion
Pulsar Functions support a variety of deployment options. You can deploy a Pulsar Function in the following ways.
```

Deployment mode | Description
:---------------|:-----------
[Local run mode](#local-run-mode) | The function runs in your local environment, for example, on your laptop.
[Cluster mode](#cluster-run-mode) | The function runs *inside of* your Pulsar cluster, on the same machines as your Pulsar [brokers](reference-terminology.md#broker).

### Local run mode

> If you run Pulsar Functions in the **local run** mode, you run it on the machine where you run commands(for example, your laptop, an [AWS EC2](https://aws.amazon.com/ec2/) instance). The following example is about the [`localrun`](reference-pulsar-admin.md#localrun) command.

Reviewer: ```suggestion
If you run a Pulsar Function in the **local run** mode, you run it on the machine where you run commands(for example, your laptop, an [AWS EC2](https://aws.amazon.com/ec2/) instance). The following example is about the [`localrun`](reference-pulsar-admin.md#localrun) command.
```

}
```

### Trigger Pulsar Functions

> You can [trigger](functions-deploying.md#triggering-pulsar-functions) Pulsar Functions running in the [cluster mode](#cluster-mode) with the [command line](#command-line-interface). After triggering Pulsar Functions, you can pass a specific value to a function and get the return value *without* creating a client, sending a message to the right input topic. Triggering is useful for, but not limited to, testing and debugging purposes.

Reviewer: ```suggestion
You can [trigger](functions-deploying.md#triggering-pulsar-functions) a Pulsar Function running in the [cluster mode](#cluster-mode) with the [command line](#command-line-interface). When triggering a Pulsar Function, you can pass a specific value to the Function and get the return value *without* creating a client. Triggering is useful for, but not limited to, testing and debugging purposes.
```
